chalk-solve/src/infer/test.rs

```diff
@@ -1,9 +1,23 @@
 #![cfg(test)]
 
-use super::unify::UnificationResult;
+use super::unify::RelationResult;
 use super::*;
 use chalk_integration::interner::ChalkIr;
 
+// We just use a vec of 20 `Invariant`, since this is zipped and no substs are
+// longer than this
+#[derive(Debug)]
+struct TestDatabase;
+impl UnificationDatabase<ChalkIr> for TestDatabase {
```

Comment:
Might be worth making `TestDatabase` stateful for each function. But that adds a bunch more to each use, so I didn't do it yet. With more tests for subtyping, it might be needed.

Reply:
So, rather it being an empty struct, it could be `variances: Variances`, and then each test can decide what the variances that should be returned are.